internal/storage/pmem/pmem.go

```diff
@@ -44,69 +50,168 @@ func mountInternal(source, target string) (err error) {
 	return nil
 }
 
-// Mount mounts the pmem device at `/dev/pmem<device>` to `target`.
+// Mount mounts the pmem device at `/dev/pmem<device>` to `target` in a basic scenario.
+// In the case when mappingInfo or/and verityInfo is/are not nil, device-mapper framework
+// is used to create linear and verity targets accordingly.
 //
 // `target` will be created. On mount failure the created `target` will be
 // automatically cleaned up.
 //
 // Note: For now the platform only supports readonly pmem that is assumed to be
 // `ext4`.
-func Mount(ctx context.Context, device uint32, target string) (err error) {
+//
+// Note: both mappingInfo and verityInfo can be non-nil at the same time, in that case
+// linear target is created first and it becomes the data/hash device for verity target.
+func Mount(ctx context.Context, device uint32, target string, mappingInfo *prot.DeviceMappingInfo, verityInfo *prot.DeviceVerityInfo) (err error) {
 	_, span := trace.StartSpan(ctx, "pmem::Mount")
 	defer span.End()
 	defer func() { oc.SetSpanStatus(span, err) }()
 
 	span.AddAttributes(
-		trace.Int64Attribute("device", int64(device)),
+		trace.Int64Attribute("deviceNumber", int64(device)),
 		trace.StringAttribute("target", target))
 
-	source := fmt.Sprintf(pMemFmt, device)
-	return mountInternal(source, target)
+	devicePath := fmt.Sprintf(pMemFmt, device)
+	if mappingInfo == nil && verityInfo == nil {
+		return mountInternal(devicePath, target)
+	}
+
+	// dm linear target has to be created first. when verity info is also present, the linear target becomes the data
+	// device instead of the original VPMem.
+	if mappingInfo != nil {
+		dmLinearName := fmt.Sprintf(linearDeviceFmt, device, mappingInfo.DeviceOffsetInBytes, mappingInfo.DeviceSizeInBytes)
+		if devicePath, err = createDMLinearTarget(ctx, devicePath, dmLinearName, target, mappingInfo); err != nil {
+			return err
+		}
+		defer func() {
+			if err != nil {
+				if err := dm.RemoveDevice(dmLinearName); err != nil {
+					log.G(ctx).WithError(err).Debugf("failed to cleanup linear target: %s", dmLinearName)
+				}
+			}
+		}()
+	}
+
+	if verityInfo != nil {
+		dmVerityName := fmt.Sprintf(verityDeviceFmt, device, verityInfo.RootDigest)
+		if devicePath, err = createDMVerityTarget(ctx, devicePath, dmVerityName, target, verityInfo); err != nil {
+			return err
+		}
+		defer func() {
+			if err != nil {
+				if err := dm.RemoveDevice(dmVerityName); err != nil {
+					log.G(ctx).WithError(err).Debugf("failed to cleanup verity target: %s", dmVerityName)
+				}
+			}
+		}()
+	}
+
+	return mountInternal(devicePath, target)
+}
+
+// createDMLinearTarget creates dm-linear target from a given `device` slot location and `mappingInfo`
+func createDMLinearTarget(ctx context.Context, devPath, devName string, target string, mappingInfo *prot.DeviceMappingInfo) (_ string, err error) {
+	if mappingInfo == nil {
+		return "", errors.New("mappingInfo cannot be nil")
+	}
```

Comment:
This can never happen right?

Reply:
You're right I believe (and for the verity info also). I always think about how we'd like to handle things like this because I feel like we run into this a lot. It's impossible we'd enter here currently but if this was ever reworked/expanded on so another time this is called wasn't guarded by mappingInfo being != nil in the first place then we're in a weird spot. The check is harmless though even if a bit redundant

Reply:
My opinion is that it appropriate to just AV in cases that should never happen. One argument for this is it removes all confusion on whether the null is actually a possibility and needs to be handled.

Reply:
What does AV stand for?

Reply:
Access violation I believe

Reply:
AKA segfault/nil point dereference

Reply:
Ah, I meant an Access Violation. In other words, go ahead and let it crash/panic/throw when it tries to deref the null we never expect vs checking for null and gracefully returning an error.

Reply:
ah, gocha! makes sense! I was pinging offline in parallel ahah

Reply:
Sounds good to me. At least from a containerd perspective and not a general best practices point of view, if we panic the cause is a little harder to debug than if we just bubbled up an error. Amit alleviated this by having panics actually get logged now though